x-pack/plugins/task_manager/server/task_claimers/strategy_mget.ts

```diff
@@ -319,69 +318,115 @@ async function searchAvailableTasks({
   removedTypes,
   excludedTypes,
   taskStore,
+  getCapacity,
   size,
   taskPartitioner,
 }: OwnershipClaimingOpts): Promise<SearchAvailableTasksResponse> {
-  const searchedTypes = Array.from(taskTypes)
-    .concat(Array.from(removedTypes))
-    .filter((type) => !excludedTypes.has(type));
-  const queryForScheduledTasks = mustBeAllOf(
-    // Task must be enabled
-    EnabledTask,
-    // a task type that's not excluded (may be removed or not)
-    OneOfTaskTypes('task.taskType', searchedTypes),
-    // Either a task with idle status and runAt <= now or
-    // status running or claiming with a retryAt <= now.
-    shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt),
-    // must have a status that isn't 'unrecognized'
-    RecognizedTask
-  );
+  const claimPartitions = buildClaimPartitions({
+    types: taskTypes,
+    excludedTypes,
+    removedTypes,
+    getCapacity,
+    definitions,
+  });
   const partitions = await taskPartitioner.getPartitions();
 
   const sort: NonNullable<SearchOpts['sort']> = getClaimSort(definitions);
-  const query = matchesClauses(
-    queryForScheduledTasks,
-    filterDownBy(InactiveTasks),
-    tasksWithPartitions(partitions)
-  );
+  const searches: SearchOpts[] = [];
+
+  // not handling removed types yet
+
+  // add search for unlimited types
+  if (claimPartitions.unlimitedTypes.length > 0) {
+    const queryForUnlimitedTasks = mustBeAllOf(
+      // Task must be enabled
+      EnabledTask,
+      // a task type that's not excluded (may be removed or not)
+      OneOfTaskTypes('task.taskType', claimPartitions.unlimitedTypes),
+      // Either a task with idle status and runAt <= now or
+      // status running or claiming with a retryAt <= now.
+      shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt),
+      // must have a status that isn't 'unrecognized'
+      RecognizedTask
+    );
+
+    const queryUnlimitedTasks = matchesClauses(
+      queryForUnlimitedTasks,
+      filterDownBy(InactiveTasks),
+      tasksWithPartitions(partitions)
+    );
+    searches.push({
+      query: queryUnlimitedTasks,
+      sort,
+      size,
+      seq_no_primary_term: true,
+    });
+  }
 
-  return await taskStore.fetch(
-    {
+  // add searches for limited types
+  for (const [type, capacity] of claimPartitions.limitedTypes) {
+    const queryForLimitedTasks = mustBeAllOf(
+      // Task must be enabled
+      EnabledTask,
+      // Specific task type
+      OneOfTaskTypes('task.taskType', [type]),
+      // Either a task with idle status and runAt <= now or
+      // status running or claiming with a retryAt <= now.
+      shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt),
+      // must have a status that isn't 'unrecognized'
+      RecognizedTask
+    );
+
+    const query = matchesClauses(queryForLimitedTasks, filterDownBy(InactiveTasks));
```

Comment:
do we need to add `tasksWithPartitions` to this clause?

Reply:
added in 842d4023a447ed91b81627faee668b0bcb2a1e16